src/core/border_router/nat64_translator.hpp

```diff
@@ -0,0 +1,185 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for the NAT64 translator.
+ *
+ */
+
+#ifndef NAT64_TRANSLATOR_HPP_
+#define NAT64_TRANSLATOR_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE
+
+#if !OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+#error "OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE is required for OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE."
+#endif
+
+#if !OPENTHREAD_CONFIG_UPTIME_ENABLE
+#error "OPENTHREAD_CONFIG_UPTIME_ENABLE is required for OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE."
+#endif
+
+#include "common/linked_list.hpp"
+#include "common/locator.hpp"
+#include "common/pool.hpp"
+#include "net/ip4_types.hpp"
+#include "net/ip6.hpp"
+
+namespace ot {
+namespace BorderRouter {
+
+class Nat64Translator : public InstanceLocator, private NonCopyable
+{
+public:
+    static constexpr uint32_t kAddressMappingIdleTimeoutMsec =
+        OPENTHREAD_CONFIG_BORDER_ROUTING_NAT64_IDLE_TIMEOUT_SECONDS * Time::kOneSecondInMsec;
+    static constexpr uint32_t kAddressMappingPoolSize = OPENTHREAD_CONFIG_BORDER_ROUTING_NAT64_MAX_MAPPINGS;
+
+    enum class Result : uint8_t
+    {
+        kForward   = 0,
+        kDrop      = 1,
+        kReplyIcmp = 2,
+    };
+
+    /**
+     * This constructor initializes the nat64.
+     *
+     */
+    explicit Nat64Translator(Instance &aInstance);
+
+    /**
+     * @brief Translates an IPv4 packet to IPv6 packet. Note the packet and packetLength might be adjusted. Note the
+     * caller should reserve at least 20 bytes before the packetHead.
+     * If the message is an IPv6 packet, Result::kForward will be returned and the message won't be modified.
+     *
+     * @param[in,out] aMessage the message to be processed.
+     *
+     * @returns Result::kForward the caller should contiue forwarding the packet.
+     * @returns Result::kDrop the caller should drop the packet silently.
+     * @returns Result::kReplyIcmp the caller should reply an ICMP packet, the buffer will be filled with the content of
+     * the ICMP packet.
+     *
+     */
+    Result HandleIncoming(Message &message);
```

Comment:
Howabout `HandleIp6Incoming`, `HandleIp6Outgoing`

Since the translator should peek the packets to check whether it is for NAT64, the name "TranslateIp4" more like it only accepts IPv4 messages and will drop all other messages.

Reply:
Thanks. I agree they are good name... short and clear 👍